Please run CI with more dmd versions

for projects that use unit-threaded that support a wide amount of compiler releases it would be handy if this projects CI was running against more versions of dmd. 

I'm running builds from dmd 2.080 to 2.096 and some of the older versions are failing. If you don't wish to support versions that far back then please at least support last a defined range of compilers and make it clear in the readme.

With dmd 2.081.1 and 2.082.1 on Ubuntu I'm seeing:

```
unit-threaded:from 1.0.15: building configuration "library"...
unit-threaded:exception 1.0.15: building configuration "library"...
unit-threaded:assertions 1.0.15: building configuration "library"...
unit-threaded:integration 1.0.15: building configuration "library"...
unit-threaded:mocks 1.0.15: building configuration "library"...
unit-threaded:property 1.0.15: building configuration "default"...
unit-threaded:runner 1.0.15: building configuration "library"...
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/runner/source/unit_threaded/runner/runner.d(102,35): Error: found `,` when expecting `)`
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/runner/source/unit_threaded/runner/runner.d(102,37): Error: found `module_` when expecting `;`
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/runner/source/unit_threaded/runner/runner.d(102,44): Error: found `,` instead of statement
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/runner/source/unit_threaded/runner/runner.d(103,34): Error: found `,` when expecting `)`
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/runner/source/unit_threaded/runner/runner.d(103,36): Error: found `".testData()"` when expecting `;` following statement
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/runner/source/unit_threaded/runner/runner.d(103,49): Error: found `)` instead of statement
/opt/hostedtoolcache/dc/dmd-2.082.1/x64/dmd2/linux/bin64/dmd failed with exit code 1.
Error: Process completed with exit code 2.
```

With dmd 2.080.1 on Ubuntu I'm seeing

```
Performing "debug" build using /opt/hostedtoolcache/dc/dmd-2.080.1/x64/dmd2/linux/bin64/dmd for x86_64.
unit-threaded:from 1.0.15: building configuration "library"...
unit-threaded:exception 1.0.15: building configuration "library"...
unit-threaded:assertions 1.0.15: building configuration "library"...
unit-threaded:integration 1.0.15: building configuration "library"...
unit-threaded:mocks 1.0.15: building configuration "library"...
unit-threaded:property 1.0.15: building configuration "default"...
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/property/source/unit_threaded/randomized/random.d(52,11): Error: statement expected to be `{ }`, not `(`
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/property/source/unit_threaded/randomized/random.d(53,9): Error: found `do` when expecting `;` following statement
../../../.dub/packages/unit-threaded-1.0.15/unit-threaded/subpackages/property/source/unit_threaded/randomized/random.d(54,5): Error: missing `do { ... }` after `in` or `out`
/opt/hostedtoolcache/dc/dmd-2.080.1/x64/dmd2/linux/bin64/dmd failed with exit code 1.
Error: Process completed with exit code 2.
```

thanks for the branch. I made a fork and took a look at getting a decent build matrix going. Seems that the current codebase should be ok on D versions 2.089 to latest (2.096). I would have made a pull request for [this branch](https://github.com/SingingBush/unit-threaded/tree/ci/multiple_compilers) but there's still the odd problem. Generally on LDC:

| Compiler      | Ubuntu | Windows | Mac |
| ----------- | ----------- | ----------- | ----------- |
| dmd-latest (2.096.1)   | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| dmd-2.095.1  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| dmd-2.094.2  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| dmd-2.093.1  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| dmd-2.092.1  | :heavy_check_mark: | :x: | :heavy_check_mark: |
| dmd-2.091.1  | :heavy_check_mark: | :x: | :heavy_check_mark: |
| dmd-2.090.1  | :heavy_check_mark: | :x: | :heavy_check_mark: |
| dmd-2.089.1  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| ldc-latest  | :x: | :heavy_check_mark: | :x: |
| ldc-25  | :x: | :heavy_check_mark: | :x: |
| ldc-24  | :x: | :heavy_check_mark: | :x: |
| ldc-23  | :x: | :heavy_check_mark: | :x: |

The ldc builds could probably be sorted easy as it just seems to be looking for dmd for some reason:

```
Regular tests
--------------------

unit_threaded.exception.UnitTestException@tests/unit_threaded/ut/integration.d(103): Could not execute `dmd /Users/runner/work/unit-threaded/unit-threaded/tmp/unit-threaded/YYYYYYILKqr3/hello.d`:
Executable file not found: dmd
----------------
```

On Windows I was seeing linker errors:

```
Issue 61
Performing "unittest" build using C:\hostedtoolcache\windows\dc\dmd-2.091.1\x64\dmd2\windows\bin\dmd.exe for x86_64.
unit-threaded:from ~ci/multiple_compilers: building configuration "library"...
unit-threaded:exception ~ci/multiple_compilers: building configuration "library"...
unit-threaded:assertions ~ci/multiple_compilers: building configuration "library"...
unit-threaded:integration ~ci/multiple_compilers: building configuration "library"...
unit-threaded:mocks ~ci/multiple_compilers: building configuration "library"...
unit-threaded:property ~ci/multiple_compilers: building configuration "default"...
unit-threaded:runner ~ci/multiple_compilers: building configuration "library"...
unit-threaded ~ci/multiple_compilers: building configuration "library"...
issue61 ~master: building configuration "application"...
Linking...
issue61.obj : fatal error LNK1179: invalid or corrupt file: duplicate COMDAT '_D3std6traits__T6fqnSymS3foo3barZ11adjustIdentFAyaZQe'
Error: linker exited with status 1179
C:\hostedtoolcache\windows\dc\dmd-2.091.1\x64\dmd2\windows\bin\dmd.exe failed with exit code 1.
Error: Process completed with exit code 1.
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -1,20 +1,38 @@
 name: CI
-on: [push, pull_request]
+
+on:
+  schedule:
+    - cron: '30 7 1 * *'
+  push:
+  pull_request:
 
 jobs:
   test:
     name: Dub Test
     strategy:
+      fail-fast: false
       matrix:
-        os: [ubuntu-20.04, windows-2019, macos-10.15]
-        dc: [dmd-2.096.0]
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        dc:
+          - dmd-latest
+          - ldc-latest
+          - dmd-2.095.1
+          - dmd-2.094.2
+          - dmd-2.093.1
+          #- dmd-2.092.1 # there are linker errors on Windows for dmd 2.090 - 2.092
+          #- dmd-2.091.1
+          #- dmd-2.090.1
+          - dmd-2.089.1 # builds fail for D versions below 2.089
+          - ldc-1.25.1 # eq to dmd v2.095.1
+          - ldc-1.24.0 # eq to dmd v2.094.1
+          - ldc-1.23.0 # eq to dmd v2.093.1
 
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
 
-      - name: Install D compiler
-        uses: dlang-community/setup-dlang@v1.0.5
+      - name: Install ${{ matrix.dc }}
+        uses: dlang-community/setup-dlang@v1.1.0
         with:
           compiler: ${{ matrix.dc }}
 
@@ -27,5 +45,22 @@ jobs:
       - name: "Windows: Run tests"
         if: runner.os == 'Windows'
         run: build\ci.bat
+  
+  coverage:
+    name: Code coverage
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Install ${{ matrix.dc }}
+        uses: dlang-community/setup-dlang@v1.1.0
+        with:
+          compiler: ${{ matrix.dc }}
+
+      - name: "Posix: Run tests"
+        if: runner.os != 'Windows'
+        run: build/ci.sh
+        env:
+          TERM: xterm
 
       - uses: codecov/codecov-action@v1

--- a/dub.json
+++ b/dub.json
@@ -33,7 +33,7 @@
 
         {
             "name": "library",
-            "dflags": ["-preview=dip25", "-preview=dip1000"],
+            "dflags": ["-preview=dip25", "-preview=dip1000"]
         },
 
         { "name": "nodips" },
@@ -45,7 +45,7 @@
             "sourcePaths": ["source", "gen/source", "tests/unit_threaded", "tests/examples/pass"],
             "importPaths": ["source"],
             "mainSourceFile": "example/example_pass.d",
-            "dflags": ["-preview=dip25", "-preview=dip1000", "-preview=dip1008", "-cov=ctfe"],
+            "dflags": ["-preview=dip25", "-preview=dip1000", "-preview=dip1008"],
             "versions": ["testing_unit_threaded"]
         },
         {

--- a/subpackages/runner/source/unit_threaded/runner/testcase.d
+++ b/subpackages/runner/source/unit_threaded/runner/testcase.d
@@ -374,20 +374,27 @@ private string localStacktraceToString(Throwable throwable, int removeExtraLines
 }
 
 unittest {
-    import std.string : splitLines;
+    import std.conv : to;
+    import std.string : splitLines, indexOf;
     import std.format : format;
 
     try throw new Exception("");
     catch (Exception exc) {
         const output = exc.localStacktraceToString(0);
-        const lines = output.splitLines.length;
+        const lines = output.splitLines;
 
         /*
          * object.Exception@subpackages/runner/source/unit_threaded/runner/testcase.d(368)
          * ----------------
          * subpackages/runner/source/unit_threaded/runner/testcase.d:368 void unit_threaded.runner.testcase [...]
          */
-        assert(lines == 3);
+        //import std.stdio : writeln;
+        //writeln("Output from local stack trace was:\n"~output);
+
+        assert(lines.length >= 3, "Expected 3 or more lines but got " ~ to!string(lines.length) ~ " :\n" ~ output);
+        assert(lines[0].indexOf("object.Exception@subpackages/runner/source/unit_threaded/runner/testcase.d") != -1);
+	    assert(lines[1].indexOf("------") != -1); // second line is a bunch of dashes
+        assert(lines[2].indexOf("testcase.d") != -1); // the third line differs accross compilers
     }
 }
 

--- a/subpackages/runner/source/unit_threaded/runner/testcase.d
+++ b/subpackages/runner/source/unit_threaded/runner/testcase.d
@@ -384,17 +384,20 @@ unittest {
         const lines = output.splitLines;
 
         /*
+         * The text of a stacktrace can differ between compilers and also paths differ between Unix and Windows.
+         * Example exception test from dmd on unix:
+         *
          * object.Exception@subpackages/runner/source/unit_threaded/runner/testcase.d(368)
          * ----------------
          * subpackages/runner/source/unit_threaded/runner/testcase.d:368 void unit_threaded.runner.testcase [...]
          */
-        //import std.stdio : writeln;
-        //writeln("Output from local stack trace was:\n"~output);
+        import std.stdio : writeln;
+        writeln("Output from local stack trace was " ~ to!string(lines.length) ~ " lines:\n"~output~"\n");
 
         assert(lines.length >= 3, "Expected 3 or more lines but got " ~ to!string(lines.length) ~ " :\n" ~ output);
-        assert(lines[0].indexOf("object.Exception@subpackages/runner/source/unit_threaded/runner/testcase.d") != -1);
+        assert(lines[0].indexOf("object.Exception@") != -1, "Line 1 of stack trace should show exception type. Was: "~lines[0]);
 	    assert(lines[1].indexOf("------") != -1); // second line is a bunch of dashes
-        assert(lines[2].indexOf("testcase.d") != -1); // the third line differs accross compilers
+        //assert(lines[2].indexOf("testcase.d") != -1); // the third line differs accross compilers and not reliable for testing
     }
 }
 

--- a/tests/unit_threaded/ut/integration.d
+++ b/tests/unit_threaded/ut/integration.d
@@ -100,7 +100,13 @@ version(Windows) {
         shouldFail("definitely_not_an_existing_command_or_executable");
 
         writeFile("hello.d", q{import std; void main() {writeln("hello");}});
-        shouldExecuteOk(["dmd", inSandboxPath("hello.d")]);
+
+        version(LDC) {
+            shouldExecuteOk(["ldc2", inSandboxPath("hello.d")]);
+        } else {
+            shouldExecuteOk(["dmd", inSandboxPath("hello.d")]);
+        }
+        
         version (Windows)
             immutable exe = "hello.exe";
         else

